projects/convai2/baselines/profilememory/eval_ppl.py

```diff
@@ -0,0 +1,166 @@
+
+    def share(self):
+        shared = super().share()
+        shared['probs'] = self.probs.copy()
+        return shared
+
+    def next_word_probability(self, observation, partial_out):
+        """Return probability distribution over next words given an input and
+        partial true output. This is used to calculate the per-word perplexity.
+        Arguments:
+        observation -- input observation dict
+        partial_out -- previous "true" words
+        Returns a dict, where each key is a word and each value is a probability
+        score for that word. Unset keys assume a probability of zero.
+        e.g.
+        {'text': 'Run test program.'}, ['hello'] => {'world': 1.0}
+        """
+        if not hasattr(self, 'last_text'):
+            self.last_text = None
+
+        if observation['text'] != self.last_text:
+            self.last_text = observation.get('text')
+            self.observe(observation)
+
+        obs = [self.observation]
+
+        if len(partial_out) > 0:
+            obs[0]['eval_labels'] = (' '.join(partial_out),)
+        else:
+            obs[0]['eval_labels'] = ['']
+
+        xs, xs_persona, ys, labels, valid_inds, cands, valid_cands, zs, eval_labels = self.batchify(obs)
+
+        self.encoder.train(mode=False)
+        self.encoder_persona.train(mode=False)
+        self.decoder.train(mode=False)
+
+        encoder_output_persona, hidden_persona, guide_indices = self._encode_persona(xs_persona, ys, False)
+        encoder_output, hidden = self._encode(xs, False)
+
+        # next we use END as an input to kick off our decoder
+        x = Variable(self.START_TENSOR)
+        xe = self.lt(x).unsqueeze(1)
+        xes = xe.expand(xe.size(0), 1, xe.size(2))
+
+        attn_mask = Variable(xs_persona.data.sum(2).ne(0), requires_grad=False)
+
+        output_lines = [[] for _ in range(1)]
+
+        # we only ever want to generate the next token
+        self.longest_label = 1
+
+        attn_w_visual_tmp = []
+
+        if zs is None:
+            rnge = 1
+        else:
+            rnge = zs.size(1)
+
+        for i in range(rnge):
+            if type(self.decoder) == torch.nn.LSTM:
+                h = hidden[0]
+            if self.attention:
+                if self.opt['personachat_attnsentlevel']:
+                    h = self.attn_h2attn(h)
+                output, attn_weights, attn_w_premask = self._apply_attention(xes, encoder_output_persona, h, attn_mask)
+            else:
+                output = xes
+
+            output, hidden = self.decoder(output, hidden)
+
+            new_output = F.dropout(output.squeeze(0), p=self.dropout, training=False)
+            if self.sharelt:
+                e = self.h2e(new_output)
+                scores = self.e2o(e)[:,1:]
+            else:
+                scores = self.h2o(new_output)
+            probs = F.softmax(scores)
+            _max_score, preds = scores.max(1)
+
+            if zs is not None:
```

Comment:
nothing in this if block is used, right? since probs is already used above. and don't need the _max_score, preds = scores.max(1) either

Reply:
yes it is used: we get a new `xes` by feeding the true token. i was using the `preds` to generate the output lines for debugging, etc. i can delete the `output_lines` code if you want but i thought it was helpful to see what the model was predicting.

Reply:
right yeah it's only for the output lines, sure we can keep if you like them